### PR TITLE
8143900: OptimizeStringConcat has an opaque dependency on Integer.sizeTable field

### PR DESCRIPTION
**Pull request by y1yang0**
Hi, can I have a review for this patch? I noticed a strange field Integer.sizeTable which is used by PhaseStringOpts, after digging into the history, I think it could be replaced by an in-place array allocation and initialization. Before it, we are fetching from Integer.sizeTable and get num of digit in integer by iterating size table, now we fetch from in-place sizeTable and get size from that. The changed IR looks like this:

image

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8143900](https://bugs.openjdk.org/browse/JDK-8143900): OptimizeStringConcat has an opaque dependency on Integer.sizeTable field

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12680/head:pull/12680` \
`$ git checkout pull/12680`

Update a local copy of the PR: \
`$ git checkout pull/12680` \
`$ git pull https://git.openjdk.org/jdk pull/12680/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12680`

View PR using the GUI difftool: \
`$ git pr show -t 12680`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12680.diff">https://git.openjdk.org/jdk/pull/12680.diff</a>

</details>

### CODE REVIEW

**Comment by y1yang0** [comment]
I updated this patch to use the algorithm identical to Integer.stringSize.

File: src/hotspot/share/opto/stringopts.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/stringopts.cpp:1163]: `Integer.stringSize() method`

**Review by vnkozlov — COMMENTED** [review]
Code looks reasonable and matches `Integer.stringSize()` method. I will test it.

**Comment by vnkozlov** [comment]
Testing results seem good. Except one strange failure I put in [JDK-8143900](https://bugs.openjdk.org/browse/JDK-8143900) comment.
You need second review.

**Comment by y1yang0** [comment]
> Testing results seem good. Except one strange failure I put in [JDK-8143900](https://bugs.openjdk.org/browse/JDK-8143900) comment. You need second review.

This seems related to https://bugs.openjdk.org/browse/JDK-8296914

**Comment by vnkozlov** [comment]
> > Testing results seem good. Except one strange failure I put in [JDK-8143900](https://bugs.openjdk.org/browse/JDK-8143900) comment. You need second review.
> 
> This seems related to https://bugs.openjdk.org/browse/JDK-8296914

[JDK-8296914 ](https://bugs.openjdk.org/browse/JDK-8296914)was closed as duplicate of still opened [JDK-8270202](https://bugs.openjdk.org/browse/JDK-8270202)

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Comment by y1yang0** [comment]
Thanks @vnkozlov @TobiHartmann for reviews.

/integrate

**Pull request closed without merging**
